src/sys/DISK.ASM

```diff
@@ -275,15 +275,33 @@ write_file:                                     ; Write file buffer to disk as f
     jmp .end                                    ; We're done!
 
 .pre_apply:
-    mov word [file_buffer_pointer], 0x0000      ; Start reading from offset 0
+    cmp byte [filename_extension], 0x20         ; Is this a directory?
+    je .check_existence                         ; If so, skip this step
+
     mov word [fsize], cx                        ; Save our file size
 
     mov si, iface_disk_applyw
     call comment_line                           ; Comment message
 
+    push word [file_buffer_segment]
+    push word [file_buffer_pointer]             ; Preserve file buffer location
+    mov cl, byte [dir_id]                       ; Preserve directory ID
+
+    cld                                         ; Copy to backup
+    call copy_fname_buffer                      ; Back up filename buffer
+
     call prompt_apply                           ; Apply filesystem changes before writing file
+
+    pop word [file_buffer_pointer]
+    pop word [file_buffer_segment]              ; Restore file buffer location
+
     jc .abort                                   ; If not authorized, abort
 
+    mov byte [dir_id], cl                       ; Otherwise, restore directory ID
+
+    std                                         ; Restore from backup
+    call copy_fname_buffer                      ; Restore original filename
+
```

Comment:
These values are preserved to allow for a call to `write_file` when applying directory changes... yeah, I really regret using an APPLY/ROLLBACK paradigm for directory operations.